File: Problems/Algorithms/103. Binary Tree Zigzag Level Order Traversal/ZigzagLevelOrderTraversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun zigzagLevelOrder(root: TreeNode?): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        
        if (root == null) {
            return result
        }
        
        val queue = ArrayDeque<TreeNode?>()
        queue.add(root)
        
        while (!queue.isEmpty()) {
            val len = queue.size
            val curr = mutableListOf<Int>()
            
            for (i in 0..len-1) {
                val node = queue.removeFirst()
                curr.add(node?.`val`!!)
                
                if (node?.left != null) {
                    queue.add(node?.left)
                }
                if (node?.right != null) {
                    queue.add(node?.right)
                }
            }
            
            result.add(curr)
        }
        
        for (i in 0..result.size-1) {
            if (i % 2 == 1) {
                result[i] = result[i].reversed()
            }
        }
        
        return result
    }
}

File: Problems/Algorithms/103. Binary Tree Zigzag Level Order Traversal/zigzag_level_order_traversal.py
#!/usr/bin/env python3

from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        if root is None:
            return result
        
        queue = [root]
        while len(queue) > 0:
            length = len(queue)
            curr = []
            
            for i in range(length):
                node = queue.pop(0)
                curr.append(node.val)
                
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)
            
            result.append(curr)
        
        for i in range(len(result)):
            if i % 2 == 1:
                result[i] = result[i][::-1]
        
        return result
